service/topology_coordinator.cc

```diff
@@ -1339,9 +1346,30 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
     future<> generate_migration_updates(utils::chunked_vector<canonical_mutation>& out, const group0_guard& guard, const migration_plan& plan) {
         if (plan.resize_plan().finalize_resize.empty() || plan.has_nodes_to_drain()) {
             // schedule tablet migration only if there are no pending resize finalisations or if the node is draining.
-            for (const tablet_migration_info& mig : plan.migrations()) {
-                co_await coroutine::maybe_yield();
-                generate_migration_update(out, guard, mig);
+
+            auto drain_failures = plan.drain_failures();
+
+            // Need to cancel decommission requests if nodes got excluded to avoid unintended rebuilds.
+            for (auto [node, req] : _topo_sm._topology.paused_requests) {
+                if (req == topology_request::leave && _topo_sm._topology.excluded_tablet_nodes.contains(node)) {
+                    drain_failures.push_back(drain_failure(locator::host_id(node.uuid()), "decommissioned node is excluded"));
+                }
+            }
+
+            // Do not schedule migrations if there are drain failures because the plan is invalid.
+            // Load allocation will change after drain is lifted, but also we may have scheduled
+            // unintended rebuilds for excluded decommissioning nodes.
+            if (!drain_failures.empty()) {
+                for (auto drain_fail : drain_failures) {
+                    co_await coroutine::maybe_yield();
+                    auto server_id = raft::server_id(drain_fail.node().uuid());
+                    _topo_sm.generate_cancel_request_update(out, _feature_service, guard, server_id, drain_fail.reason());
+                }
+            } else {
```

Comment:
Since we omit all migrations if there are drain failures, `make_plan` can return early if there are any `drain_failures`

Reply:
Yes, that already mostly happens. We break out of inter-node plan. Only intra-node plan may be generated.

We could optimize it, but it's a rare event.